rust/otap-dataflow/crates/engine/src/exporter.rs

```diff
@@ -16,48 +17,68 @@
 //!
 //! 1. The exporter is instantiated and configured
 //! 2. The `start` method is called, which begins the exporter's operation
-//! 3. The exporter processes both internal control messages and pipeline data
-//! 4. The exporter shuts down when it receives a `Shutdown` control message or encounters a fatal error
+//! 3. The exporter processes both internal control messages and pipeline data (pdata)
+//! 4. The exporter shuts down when it receives a `Shutdown` control message or encounters a fatal
+//!    error
 //!
 //! # Thread Safety
 //!
-//! Note that this trait uses `#[async_trait(?Send)]`, meaning implementations
-//! are not required to be thread-safe. To ensure scalability, the pipeline engine will start
-//! multiple instances of the same pipeline in parallel, each with its own exporter instance.
+//! Note that this trait uses `#[async_trait(?Send)]`, meaning implementations are not required to
+//! be thread-safe. If you need to implement an exporter that requires `Send`, you can use the
+//! [`SendEffectHandler`] type. The default effect handler is `!Send` (see
+//! [`NotSendEffectHandler`]).
+//!
+//! # Scalability
+//!
+//! To ensure scalability, the pipeline engine will start multiple instances of the same pipeline
+//! in parallel on different cores, each with its own exporter instance.
 
-use crate::NodeName;
+use crate::config::ExporterConfig;
 use crate::error::Error;
 use crate::message::{ControlMsg, Message};
 use async_trait::async_trait;
 use otap_df_channel::error::RecvError;
 use otap_df_channel::mpsc;
+use std::marker::PhantomData;
+use std::pin::Pin;
 use std::rc::Rc;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::time::{Instant, Sleep, sleep_until};
 
 /// A trait for egress exporters.
+///
+/// Note: The default effect handler is `!Send` (see [`NotSendEffectHandler`]).
 #[async_trait(?Send)]
-pub trait Exporter {
-    /// The type of messages handled by the exporter.
-    type PData;
-
+pub trait Exporter<PData, EF = NotSendEffectHandler<PData>>
```

Comment:
Nice! I like the idea of using `PData` as a generic type parameter instead of an associated type. It would allow me to have different implementations (OTLP and OTAP) for the same exporter type.